File: backend/src/models/Skill.js
const mongoose = require('mongoose');

const SkillSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
  },
  category: {
    _id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'SkillCategory',
      required: true
    },
    name: {
      type: String,
      required: true
    }
  },
  description: {
    type: String,
    required: true
  },
  relatedSkills: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Skill'
  }],
  proficiency: {
    type: String,
    enum: ['Beginner', 'Intermediate', 'Advanced'],
    default: 'Beginner'
  },
  difficultyLevel: {
    type: String,
    enum: ['Beginner', 'Intermediate', 'Advanced', 'Expert'],
    default: 'Beginner'
  },
  roadmap: [{
    subskillId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Skill',
      required: true
    },
    description: String,
    required: {
      type: Boolean,
      default: true
    },
    order: {
      type: Number,
      default: 0
    }
  }],
  resources: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Resource'
  }],
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  createdAt: {
    type: Date,
    default: Date.now
  }
});

module.exports = mongoose.model('Skill', SkillSchema);

File: frontend/lib/models/skill.dart
import 'resource_model.dart';

class Skill {
  final String id;
  final String name;
  final String category;
  final String description;
  final String difficultyLevel;
  final List<String> relatedSkills;
  final List<Resource> resources;
  final String? createdBy;
  final String? proficiency;

  Skill({
    required this.id,
    required this.name,
    required this.category,
    required this.description,
    required this.difficultyLevel,
    required this.relatedSkills,
    required this.resources,
    this.createdBy,
    this.proficiency,
  });

  factory Skill.fromJson(Map<String, dynamic> json) {
    return Skill(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      category: json['category'] ?? '',
      description: json['description'] ?? '',
      difficultyLevel: json['difficultyLevel'] ?? 'beginner',
      relatedSkills: List<String>.from(json['relatedSkills'] ?? []),
      resources: (json['resources'] as List<dynamic>?)
              ?.map((r) => Resource.fromJson(r))
              .toList() ??
          [],
      createdBy: json['createdBy'],
      proficiency: json['proficiency'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'category': category,
      'description': description,
      'difficultyLevel': difficultyLevel,
      'relatedSkills': relatedSkills,
      'resources': resources.map((r) => r.toJson()).toList(),
      'createdBy': createdBy,
      'proficiency': proficiency,
    };
  }

  Skill copyWith({
    String? id,
    String? name,
    String? category,
    String? description,
    String? difficultyLevel,
    List<String>? relatedSkills,
    List<Resource>? resources,
    String? createdBy,
    String? proficiency,
  }) {
    return Skill(
      id: id ?? this.id,
      name: name ?? this.name,
      category: category ?? this.category,
      description: description ?? this.description,
      difficultyLevel: difficultyLevel ?? this.difficultyLevel,
      relatedSkills: relatedSkills ?? this.relatedSkills,
      resources: resources ?? this.resources,
      createdBy: createdBy ?? this.createdBy,
      proficiency: proficiency ?? this.proficiency,
    );
  }
}
